pkg/ovs/ovsctl/appctl_test.go

```diff
@@ -110,3 +125,128 @@ IPv6 ND Extension: No`,
 		})
 	}
 }
+
+func TestTrace(t *testing.T) {
+	assert := assert.New(t)
+	srcMAC, _ := net.ParseMAC("11:11:11:11:11:11")
+	dstMAC, _ := net.ParseMAC("11:11:11:11:11:22")
+	srcIP := net.ParseIP("1.1.1.2")
+	dstIP := net.ParseIP("2.2.2.2")
+	tc := []struct {
+		name          string
+		req           *TracingRequest
+		expectedFlow  string
+		expectedError string
+	}{
+		{
+			name: "Trace IP Packet-success",
+			req: &TracingRequest{
+				InPort:              "9",
+				SrcIP:               srcIP,
+				DstIP:               dstIP,
+				SrcMAC:              srcMAC,
+				DstMAC:              dstMAC,
+				Flow:                "in_port=3,tcp,tcp_dst=22,",
+				AllowOverrideInPort: true,
+			},
+			expectedFlow: "dl_src=11:11:11:11:11:11,dl_dst=11:11:11:11:11:22,in_port=3,tcp,tcp_dst=22,,nw_ttl=64,nw_src=1.1.1.2,nw_dst=2.2.2.2,",
+		},
+		{
+			name: "Trace Non-IP Packet-success",
+			req: &TracingRequest{
+				InPort: "9",
+				SrcMAC: srcMAC,
+				DstMAC: dstMAC,
+				Flow:   "table=10, n_packets=0, n_bytes=0, priority=200,arp,actions=resubmit(,20)",
+			},
+			expectedFlow: "in_port=9,dl_src=11:11:11:11:11:11,dl_dst=11:11:11:11:11:22,table=10, n_packets=0, n_bytes=0, priority=200,arp,actions=resubmit(,20),",
+		},
+		{
+			name: "no-ops due to source and destination IP error for non-IP packet",
+			req: &TracingRequest{
+				InPort: "9",
+				SrcIP:  srcIP,
+				Flow:   "table=10, n_packets=0, n_bytes=0, priority=200,arp,actions=resubmit(,20)",
+			},
+			expectedError: "source and destination must not be specified for non-IP packet",
+		},
+		{
+			name: "no-ops due to duplicated source IP error",
+			req: &TracingRequest{
+				InPort: "9",
+				SrcIP:  srcIP,
+				Flow:   "nw_src=1.1.1.2,tcp,tcp_dst=22,",
+			},
+			expectedError: "duplicated 'nw_src' in flow",
+		},
+		{
+			name: "no-ops due to duplicated destination IP error",
+			req: &TracingRequest{
+				InPort: "9",
+				DstIP:  dstIP,
+				Flow:   "nw_dst=2.2.2.2,tcp,tcp_dst=22,",
+			},
+			expectedError: "duplicated 'nw_dst' in flow",
+		},
+		{
+			name: "no-ops due to duplicated in_port error",
+			req: &TracingRequest{
+				InPort:              "9",
+				Flow:                "in_port=3,tcp,tcp_dst=22,",
+				AllowOverrideInPort: false,
+			},
+			expectedError: "duplicated 'in_port' in flow",
+		},
+	}
+	for _, tt := range tc {
+		t.Run(tt.name, func(t *testing.T) {
+			client := &ovsCtlClient{
+				bridge: "br-int",
+				runAppCtl: func(cmd string, needsBridge bool, args ...string) ([]byte, *ExecError) {
+					return []byte(args[0]), nil
+				},
+			}
+			out, err := client.Trace(tt.req)
+			if tt.expectedError != "" {
+				assert.EqualError(err, tt.expectedError)
+			} else {
+				assert.NoError(err)
+				assert.Equal(tt.expectedFlow, out)
+			}
+		})
+	}
+}
+
+func TestDeleteDPInterface(t *testing.T) {
```

Comment:
this feels like a very hack test right now. It's hard to figure what's going on or what you are testing.
IMO, you should use mocks for `runAppCtl` and `runOVSCtl` so you can validate the different calls to these functions properly.

Reply:
@antoninbas for using the mocj functions in mock_ovsctl.go file in https://github.com/antrea-io/antrea/tree/main/pkg/ovs/ovsctl/testing I am getting cycling import error ...For resolving it I'll need to make changes to mock_ovsctl.go itself.Can you sugest whether I can do that? And do i need to use the mocks for other tests in ofctl_test.go and appctl_test.go or the method I have used is fine?

Reply:
These import cycles when mocks are generated in the `testing` sub-package can generally be addressed by generating the mocks in the same package as the interface definition and tests instead.
This is achieved by editing `hack/update-codegen-dockerized.sh` and changing the destination directory for the mocks. For example, in your case, this change could resolve the issue:
```
diff --git a/hack/update-codegen-dockerized.sh b/hack/update-codegen-dockerized.sh
index d4b0b72f4..7445602ed 100755
--- a/hack/update-codegen-dockerized.sh
+++ b/hack/update-codegen-dockerized.sh
@@ -63,7 +63,7 @@ MOCKGEN_TARGETS=(
   "pkg/ipfix IPFIXExportingProcess,IPFIXRegistry,IPFIXCollectingProcess,IPFIXAggregationProcess testing"
   "pkg/ovs/openflow Bridge,Table,Flow,Action,CTAction,FlowBuilder,Group,BucketBuilder,PacketOutBuilder testing"
   "pkg/ovs/ovsconfig OVSBridgeClient testing"
-  "pkg/ovs/ovsctl OVSCtlClient testing"
+  "pkg/ovs/ovsctl OVSCtlClient ."
   "pkg/querier AgentNetworkPolicyInfoQuerier,AgentMulticastInfoQuerier testing"
   "pkg/flowaggregator/querier FlowAggregatorQuerier testing"
   "third_party/proxy Provider testing"
```

I cannot guarantee that this is exactly what you need, I didn't look into the details of your specific issue.

Reply:
Will update this on monday @antoninbas , Does this need to be done for all test functions or only TestDeleteDPInterface

Reply:
When you use mocks, I expect the code will look different as `runAppCtl and `runOVSCtl` will be replaced by an interface. So I expect that this will have an impact on all test functions, not just `TestDeleteDPInterface`?